docs/reference/modules/discovery/fault-detection.asciidoc

```diff
@@ -31,3 +31,297 @@ The elected master node will also remove nodes from the cluster if nodes are una
+time.
+
+To troubleshoot a cluster in this state, first ensure that the cluster has a
+<<modules-discovery-troubleshooting,stable master>>, and then focus on the
+nodes unexpectedly leaving the cluster ahead of all other issues. Until the
+cluster has a stable master and stable membership there is no point in trying
+to understand or resolve any other problems in your cluster.
+
+Diagnostics and statistics are usually not useful in an unstable cluster, since
+they offer a single snapshot of the state of the cluster. Instead, look at the
+cluster logs to see the pattern of behaviour over time. Focus particularly on
+the logs from the elected master. When a node leaves the cluster, the elected
+master logs a message like this (with line breaks added for clarity):
+
+```
+[2020-01-29T10:20:35,509][INFO ][o.e.c.s.MasterService    ]
+    [es-master-1] node-left[
+        {es-master-2}{1vSx_i__QCOjVyfA9kYEkg}{q78rb3MsRDyCcweeygxeiA}{10.244.35.101}{10.244.35.101:9300}{m}
+            disconnected,
+        {es-data2-5}{KZMiSjy3QRiJevGQokB-zQ}{stIE-HcITQqLuCbAEgbCiA}{10.244.2.213}{10.244.2.213:9300}{d}
+            disconnected],
+        term: 469, version: 446673, ...
+```
+
+This message says that the `MasterService` on the elected master
+(`es-master-1`) is processing a `node-left` task, and it lists the nodes that
+are being removed and the reasons for their removal. You will also see
+similar-sounding messages on all other nodes like this:
+
+```
+[2020-01-29T10:20:48,985][INFO ][o.e.c.s.ClusterApplierService]
+    [es-master-1] removed {
+        {es-master-2}{1vSx_i__QCOjVyfA9kYEkg}{q78rb3MsRDyCcweeygxeiA}{10.244.35.101}{10.244.35.101:9300}{m}{xpack.installed=true},
+        {es-data2-5}{KZMiSjy3QRiJevGQokB-zQ}{stIE-HcITQqLuCbAEgbCiA}{10.244.2.213}{10.244.2.213:9300}{d}{xpack.installed=true}},
+    term: 469, version: 446676, reason: Publication{term=469, version=446676}
+```
+
+{es} emits this message on every node when applying a cluster state that
+removes one or more nodes from the cluster. There's less useful information
+here, and it's better to focus on the one from the `MasterService` which is
+only emitted on the elected master. If you don't see the messages from the
+`MasterService` then check you're looking at the logs from the elected master
+covering the correct time period and that logging is enabled at `INFO` level.
+
+Nodes will also log a message containing the string `master node changed`
+whenever they start or stop following the elected master. You can use these
+messages to determine each node's view of the state of the master over time.
+
+If a node restarts then it will leave the cluster and then join the cluster
+again. When it rejoins, the `MasterService` will log that it is processing a
+`node-join` task. You can tell from the master logs that the node was restarted
+because the `node-join` message will indicate that it is
+`joining after restart`. In older versions you can also determine that a node
```

Comment:
Do we need to include info about older versions? If so, we may want to note the actual version date.
```suggestion
`joining after restart`. In older {es} versions, you can also determine that a node
```

Reply:
I think so at least for now - I'm not going to backport these docs to unmaintained branches but I think they'll still be used with older versions.